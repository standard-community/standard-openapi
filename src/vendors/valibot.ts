import { toJsonSchema } from "@standard-community/standard-json";
import type { ConversionConfig } from "@valibot/to-json-schema";
import type { OpenAPIV3_1 } from "openapi-types";
import { convertToOpenAPISchema } from "./convert.js";
import type { OpenAPIMetadata, ToOpenAPISchemaFn } from "./utils.js";

export default function getToOpenAPISchemaFn(): ToOpenAPISchemaFn {
  return (schema, context) =>
    toJsonSchema(schema, {
      // @ts-expect-error
      overrideAction: ({ valibotAction, jsonSchema }) => {
        const _jsonSchema = convertToOpenAPISchema(jsonSchema, context);
        if (
          valibotAction.kind === "metadata" &&
          valibotAction.type === "metadata" &&
          !("$ref" in _jsonSchema)
        ) {
          // @ts-expect-error
          const metadata = valibotAction.metadata as OpenAPIMetadata;

          // Add example or examples to the JSON Schema
          if (metadata.example !== undefined) {
            _jsonSchema.example = metadata.example;
          }
          if (metadata.examples && metadata.examples.length > 0) {
            _jsonSchema.examples = metadata.examples;
          }

          // If a ref is provided, use it to create a $ref in the OpenAPI components
          if (metadata.ref) {
            context.components.schemas = {
              ...context.components.schemas,
              [metadata.ref]: _jsonSchema,
            };
            return {
              $ref: `#/components/schemas/${metadata.ref}`,
            };
          }
        }

        return _jsonSchema;
      },
      ...context.options,
    } satisfies ConversionConfig) as OpenAPIV3_1.SchemaObject;
}
