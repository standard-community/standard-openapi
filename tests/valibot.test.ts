import * as v from "valibot";
import { describe, expect, it } from "vitest";

import { toOpenAPISchema } from "~/index.js";

describe("valibot", () => {
  it("basic", async () => {
    const schema = v.object({
      myString: v.string(),
      myUnion: v.union([v.number(), v.boolean()]),
    });

    const specs = await toOpenAPISchema(schema);
    expect(specs).toMatchSnapshot();
  });

  it("basic with pipe", async () => {
    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description("My neat object schema"),
    );

    const specs = await toOpenAPISchema(schema);
    expect(specs).toMatchSnapshot();
  });

  it("with metadata", async () => {
    const schema = v.pipe(
      v.object({
        myString: v.string(),
        myUnion: v.union([v.number(), v.boolean()]),
      }),
      v.description("My neat object schema"),
      v.metadata({
        ref: "MyNeatObjectSchema",
      }),
    );

    const specs = await toOpenAPISchema(schema);
    expect(specs).toMatchSnapshot();
  });
});
