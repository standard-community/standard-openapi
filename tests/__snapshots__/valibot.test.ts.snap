// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`valibot > basic 1`] = `
{
  "components": undefined,
  "schema": {
    "properties": {
      "myString": {
        "type": "string",
      },
      "myUnion": {
        "anyOf": [
          {
            "type": "number",
          },
          {
            "type": "boolean",
          },
        ],
      },
    },
    "required": [
      "myString",
      "myUnion",
    ],
    "type": "object",
  },
}
`;

exports[`valibot > basic with pipe 1`] = `
{
  "components": undefined,
  "schema": {
    "description": "My neat object schema",
    "properties": {
      "myString": {
        "type": "string",
      },
      "myUnion": {
        "anyOf": [
          {
            "type": "number",
          },
          {
            "type": "boolean",
          },
        ],
      },
    },
    "required": [
      "myString",
      "myUnion",
    ],
    "type": "object",
  },
}
`;

exports[`valibot > with metadata 1`] = `
{
  "components": {
    "schemas": {
      "MyNeatObjectSchema": {
        "description": "My neat object schema",
        "properties": {
          "myString": {
            "type": "string",
          },
          "myUnion": {
            "anyOf": [
              {
                "type": "number",
              },
              {
                "type": "boolean",
              },
            ],
          },
        },
        "required": [
          "myString",
          "myUnion",
        ],
        "type": "object",
      },
    },
  },
  "schema": {
    "$ref": "#/components/schemas/MyNeatObjectSchema",
  },
}
`;
